tree/treeplayer/src/TTreeReader.cxx

```diff
@@ -397,6 +385,78 @@ bool TTreeReader::SetProxies() {
+{
+   if (!fWarnAboutLongerFriends)
+      return;
+   if (!fTree)
+      return;
+   // Make sure all proxies are set, as in certain situations we might get to this
+   // point without having set the proxies first. For example, when processing a
+   // TChain and calling `SetEntry(N)` with N beyond the real number of entries.
+   // If the proxies can't be set return from the function and give up the warning.
+   if (!fProxiesSet && !SetProxies())
+      return;
+
+   // If we are stopping the reading because we reached the last entry specified
+   // explicitly via SetEntriesRange, do not bother the user with a warning
+   if (fEntry == fEndEntry)
+      return;
+
+   const std::string mainTreeName =
+      dynamic_cast<const TChain *>(fTree) ? fTree->GetName() : ROOT::Internal::TreeUtils::GetTreeFullPaths(*fTree)[0];
+
+   const auto *friendsList = fTree->GetListOfFriends();
+   if (!friendsList || friendsList->GetEntries() == 0)
+      return;
+
+   for (decltype(fFriendProxies.size()) idx = 0; idx < fFriendProxies.size(); idx++) {
+      const auto &fp = fFriendProxies[idx];
+      // In case the friend is indexed it may very well be that it has a different number of events
+      // e.g. the friend contains information about luminosity block and
+      // all the entries in the main tree are from the same luminosity block
+      if (fp->HasIndex())
+         continue;
+      const auto *frTree = fp->GetDirector()->GetTree();
+      if (!frTree)
+         continue;
+
+      // Need to retrieve the real friend tree from the main tree
+      // because the current friend proxy points to the tree it is currently processing
+      // i.e. the current tree in a chain in case the real friend is a TChain
+      auto frEl = static_cast<TFriendElement *>(friendsList->At(idx));
+      if (!frEl)
+         continue;
+      auto frTreeFromMain = frEl->GetTree();
+      if (!frTreeFromMain)
+         continue;
+      const auto frNTrees = [&frTreeFromMain]() {
+         if (auto frChain = dynamic_cast<TChain *>(frTreeFromMain))
+            return frChain->GetNtrees();
+         else
+            return 1;
+      }();
+      const auto frNTreesSoFar = frTreeFromMain->GetTreeNumber() + 1;
+      // We are looking for the situation where there are actually still more
+      // entries to read in the friend. So we check that either:
+      // * The current entry being read is strictly below the number of entries in the friend dataset.
+      // * There are still more files to read in the friend dataset (case of TChain where we reach the boundary of the
+      //   current friend file).
+      // In any other case we continue to the next friend.
+      // Note that frTree->GetChainOffset() works in this case because the friend proxy
+      // points to the current tree in its chain, never to the actual main tree friend
+      if (fEntry >= frTree->GetChainOffset() + frTree->GetEntriesFast() && !(frNTrees > frNTreesSoFar))
```

Comment:
```suggestion
      if (fEntry >= frTreeFromMain->GetEntriesFast())
```
At this point in the code flow we know that `fEntry` is equal to `fTree->GetEntries()` because we are at the end of the main chain/tree.  We also know that either `frNTrees > frNTreesSoFar` and then `frTreeFromMain->GetEntriesFast()` is equal to `TTree::kMaxEntries` and if `frNTrees == frNTreesSoFar` then `frTreeFromMain->GetEntriesFast()` return the total number of entries in the friend chain.